### PR TITLE
[Enhance] Add type hints in /ops:

### PR DESCRIPTION
**Pull request by WINDSKY45**
## Modification

[Enhance] Add type hints in /ops:
`fused_bias_leakyrelu.py`, 'gather_points.py`, `group_points.py`.
There is no need to add type hints in `furthest_point_sample.py` and
`info.py`.
As for `focal_loss.py`, please see #1994.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/ops/group_points.py [hunk 1]
```diff
@@ -37,15 +37,15 @@ class QueryAndGroup(nn.Module):
...
-                 min_radius=0,
-                 use_xyz=True,
-                 return_grouped_xyz=False,
-                 normalize_xyz=False,
-                 uniform_sample=False,
-                 return_unique_cnt=False,
-                 return_grouped_idx=False):
+                 max_radius: float,
+                 sample_num: int,
+                 min_radius: float = 0,
```
Comment (HAOCHENYE) [inline mmcv/ops/group_points.py:42]:
```suggestion
                 min_radius: float = 0.,
```

File: mmcv/ops/group_points.py [hunk 2]
```diff
@@ -210,8 +215,7 @@ def forward(ctx, features: torch.Tensor,
         return output
 
     @staticmethod
-    def backward(ctx,
-                 grad_out: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    def backward(ctx, grad_out: torch.Tensor) -> Tuple[torch.Tensor, None]:
```
Comment (HAOCHENYE) [inline mmcv/ops/group_points.py:218]: Why return type is `Tuple[torch.Tensor, None]`?

Comment (WINDSKY45) [reply mmcv/ops/group_points.py:218]: The return is `grad_features and None`.

File: mmcv/ops/merge_cells.py [hunk 3]
```diff
@@ -150,7 +154,7 @@ def _binary_op(self, x1, x2):
 
 class GlobalPoolingCell(BaseMergeCell):
 
-    def __init__(self, in_channels=None, out_channels=None, **kwargs):
+    def __init__(self, in_channels: int, out_channels: int, **kwargs):
```
Comment (zhouzaida) [inline mmcv/ops/merge_cells.py:157]:
```suggestion
    def __init__(self, in_channels: Optional[int] = None, out_channels: Optional[int] = None, **kwargs):
```

File: mmcv/ops/gather_points.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def forward(ctx, features: torch.Tensor,
         return output
 
     @staticmethod
-    def backward(ctx, grad_out):
+    def backward(ctx, grad_out: torch.Tensor) -> tuple:
```
Comment (teamwong111) [inline mmcv/ops/gather_points.py:40]:
```suggestion
    def backward(ctx, grad_out: torch.Tensor) -> Tuple[torch.Tensor, None]:
```

File: mmcv/ops/multi_scale_deform_attn.py [hunk 5]
```diff
@@ -20,8 +23,10 @@
 class MultiScaleDeformableAttnFunction(Function):
 
     @staticmethod
-    def forward(ctx, value, value_spatial_shapes, value_level_start_index,
-                sampling_locations, attention_weights, im2col_step):
+    def forward(ctx, value: torch.Tensor, value_spatial_shapes: torch.Tensor,
+                value_level_start_index, sampling_locations: torch.Tensor,
```
Comment (teamwong111) [inline mmcv/ops/multi_scale_deform_attn.py:27]:
```suggestion
                value_level_start_index: torch.Tensor, sampling_locations: torch.Tensor,
```

File: mmcv/ops/masked_conv.py [hunk 6]
```diff
@@ -27,7 +30,13 @@ def symbolic(g, features, mask, weight, bias, padding, stride):
             stride_i=stride)
 
     @staticmethod
-    def forward(ctx, features, mask, weight, bias, padding=0, stride=1):
+    def forward(ctx,
+                features: torch.Tensor,
+                mask: torch.Tensor,
+                weight: torch.Tensor,
+                bias: torch.Tensor,
```
Comment (teamwong111) [inline mmcv/ops/masked_conv.py:37]:
```suggestion
                weight: nn.Parameter,
                bias: nn.Parameter,
```

File: mmcv/ops/merge_cells.py [hunk 7]
```diff
@@ -42,18 +43,18 @@ class BaseMergeCell(nn.Module):
     """
 
     def __init__(self,
-                 fused_channels=256,
-                 out_channels=256,
-                 with_out_conv=True,
-                 out_conv_cfg=dict(
+                 fused_channels: Optional[int] = 256,
+                 out_channels: Optional[int] = 256,
```
Comment (HAOCHENYE) [inline mmcv/ops/merge_cells.py:47]: `Optional` is only used when the default value is None,

Comment (WINDSKY45) [reply mmcv/ops/merge_cells.py:47]: >  if an explicit value of None is allowed, the use of Optional is appropriate, whether the argument is optional or not.

refers to https://docs.python.org/3/library/typing.html#typing.Optional

Comment (teamwong111) [reply mmcv/ops/merge_cells.py:47]: > Optional[X] is equivalent to X | None (or Union[X, None]).
refers to https://docs.python.org/3/library/typing.html#typing.Optional
Your understanding may be slightly biased.

File: mmcv/ops/merge_cells.py [hunk 8]
```diff
@@ -150,7 +154,10 @@ def _binary_op(self, x1, x2):
 
 class GlobalPoolingCell(BaseMergeCell):
 
-    def __init__(self, in_channels=None, out_channels=None, **kwargs):
+    def __init__(self,
+                 in_channels: Optional[int] = None,
+                 out_channels: Optional[int] = None,
```
Comment (teamwong111) [inline mmcv/ops/merge_cells.py:159]:
```suggestion
                 in_channels: int,
                 out_channels: int,
```

The default value can be removed, which should be a bug. @jshilong have a look.

File: mmcv/ops/masked_conv.py [hunk 9]
```diff
@@ -16,7 +17,9 @@
 class MaskedConv2dFunction(Function):
 
     @staticmethod
-    def symbolic(g, features, mask, weight, bias, padding, stride):
+    def symbolic(g, features: torch.Tensor, mask: torch.Tensor,
+                 weight: torch.nn.Parameter, bias: torch.nn.Parameter,
+                 padding: int, stride: int):
```
Comment (zhouzaida) [inline mmcv/ops/masked_conv.py:22]:
```suggestion
    def symbolic(g, features, mask, weight, bias, padding, stride):
```

File: mmcv/ops/modulated_deform_conv.py [hunk 10]
```diff
@@ -19,8 +20,10 @@
 class ModulatedDeformConv2dFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, offset, mask, weight, bias, stride, padding,
-                 dilation, groups, deform_groups):
+    def symbolic(g, input: torch.Tensor, offset: torch.Tensor,
+                 mask: torch.Tensor, weight: nn.Parameter, bias: nn.Parameter,
+                 stride: int, padding: int, dilation: int, groups: int,
+                 deform_groups: int):
```
Comment (zhouzaida) [inline mmcv/ops/modulated_deform_conv.py:26]:
```suggestion
    def symbolic(g, input, offset, mask, weight, bias, stride, padding,
                 dilation, groups, deform_groups):
```

File: mmcv/ops/merge_cells.py (hunk 7 above)
Comment (zhouzaida) [reply mmcv/ops/merge_cells.py:47]: fused_channels or out_channels can be None when with_out_conv is False so Optional[int] is correct.

I create a [pr](https://github.com/open-mmlab/mmcv/pull/2062) to ensure fused_channels or out_channels can not be None when with_out_conv is True.

File: mmcv/ops/merge_cells.py (hunk 8 above)
Comment (zhouzaida) [reply mmcv/ops/merge_cells.py:159]: in_channels or out_channelscan be None when with_out_conv is False so Optional[int] is correct.
I create a [pr](https://github.com/open-mmlab/mmcv/pull/2062) to ensure in_channels or out_channels can not be None when with_out_conv is True.

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**